retry.go

```diff
@@ -109,20 +120,30 @@ func (r *Retry) postCheck(err error) bool {
 
 // Attempts sets the maximum amount of retry attempts
 // before the current error is returned.
-func (r *Retry) Attempts(n int) *Retry {
-	var iterations int
-	r.appendPreCondition(func() bool {
-		ok := iterations < n
-		iterations++
-		return ok
+// It will panic if maxAttempts is negative or 0.
```

Comment:
Should we take a uint and allow 0? I think we should avoid introducing runtime panics as much as possible :thinking:

Reply:
I can't see a reason why anyone would want to send a 0. To me, that seems highly indicate of an exceptional error.

Reply:
I'm thinking of this as a fancy for loop. It's just a control structure - doing something 0 times might not be useful but it is still a completely valid program. imo panics should indicate an invalid / unrepresentable state (like retrying -30 times).

Reply:
Fair enough, I'll change it to allow 0 then.

In terms of changing it to a `uint`, I'm not sure. It'd make using Attempts a bit more ugly cause most numbers in go are ints so you'll probably often need a cast.

Reply:
And if we go with the loop analogy, Go's for loops use ints usually so maybe Attempts should too?

Reply:
And then the thing about casts is that you can cast a negative integer to a uint, which would bypass the entire point of using the uint in the API in the first place.

Reply:
For loops are much more general though, there's no impetus to use a specific type except convention. i.e.,

```Go
for i := ""; len(i) < n; i += " " { }
```

Whereas retry specializes to natural numbers.

> And then the thing about casts is that you can cast a negative integer to a uint

imo, that wouldn't be retry's problem, it's out of scope entirely. a nice benefit of using a uint is that overflowed constants will be caught by the compiler automatically:

```Go
func main() {
	fmt.Println(uint(-1))
}

prog.go:8:18: constant -1 overflows uint
```

Reply:
> Whereas retry specializes to natural numbers.

Fair enough.

> imo, that wouldn't be retry's problem, it's out of scope entirely. a nice benefit of using a uint is that overflowed constants will be caught by the compiler automatically:

If someone ever tries to run Attempts with 0 (or a negative number), it's more likely they are passing in a variable configured by someone else. If that variable is an int and negative which is possible, it will casted into a very big uint. It's possible the developer will not notice this and the retry will effectively be a infinite with no attempt limits. If we instead opt for the `int` type, we can panic in such cases and inform the dev quickly that something is wrong.

Reply:
That's a great point.

Reply:
Another possibility is that instead of panicing, we return an error when the consumer calls Run() on the retry. I think that'd be a bit cleaner.

Reply:
There already is an error being returned if the retry doesn't run at least once. While I don't think the panic here is too dangerous, we may as well return the error since it doesn't complicate the API.

Reply:
What do you think about removing that error? Or maybe exposing it as a global variable?

Reply:
Why would we remove it? Why should we export it? It's only returned for programmer mistakes.

Reply:
Yea, never mind, you're right.